pkg/trie/triedb/node.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package triedb
+
+import (
+	"bytes"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/trie/db"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb/codec"
+)
+
+type Value interface {
+	getHash() common.Hash
+}
+
+type (
+	Inline struct {
+		Data []byte
+	}
+
+	ValueRef struct {
+		hash common.Hash
+	}
+
+	NewValueRef struct {
+		hash *common.Hash
+		Data []byte
+	}
+)
+
+func (Inline) getHash() common.Hash      { return common.EmptyHash }
+func (vr ValueRef) getHash() common.Hash { return vr.hash }
+func (vr NewValueRef) getHash() common.Hash {
+	if vr.hash == nil {
+		return common.EmptyHash
+	}
+
+	return *vr.hash
+}
+
+func NewValue(data []byte, threshold int) Value {
+	if len(data) >= threshold {
+		return NewValueRef{Data: data}
+	}
+
+	return Inline{Data: data}
+}
+
+func NewFromEncoded(encodedValue codec.NodeValue) Value {
+	switch encoded := encodedValue.(type) {
+	case codec.InlineValue:
+		return Inline{Data: encoded.Data}
+	case codec.HashedValue:
+		return ValueRef{hash: common.NewHash(encoded.Data)}
+	}
+
+	return nil
+}
+
+func InMemoryFetchedValue(value Value, prefix []byte, db db.DBGetter, fullKey []byte) ([]byte, error) {
+	switch v := value.(type) {
+	case Inline:
+		return v.Data, nil
+	case NewValueRef:
+		return v.Data, nil
+	case ValueRef:
+		prefixedKey := bytes.Join([][]byte{prefix, v.hash.ToBytes()}, nil)
+		value, err := db.Get(prefixedKey)
+		if err != nil {
+			return nil, err
+		}
+		if value != nil {
+			return value, nil
+		}
+		return value, ErrIncompleteDB
+	default:
+		panic("unreachable")
+	}
+}
+
+var EmptyNode = make([]byte, 0)
+var HashedNullNode = common.MustBlake2bHash(EmptyNode)
```

Comment:
don't think these need to be exported.
```suggestion
var emptyNode = make([]byte, 0)
var hashedNullNode = common.MustBlake2bHash(EmptyNode)
```